### DISCUSSION TITLE
How do I use two schema.prisma files and avoid `exports is not defined`?

### DISCUSSION THREAD

**Discussion by Freakspot** [prisma/prisma#20200] [Q&A] [answered]
Hey,

I have been fighting the `exports is not defined` error that has already been discussed in https://github.com/prisma/prisma/discussions/19838 and https://github.com/prisma/prisma/issues/2443. Unfortunately, I have not managed to fix my issue with the provided solutions. The issue started with me changing the `output` of my schema, because I wanted to accommodate for a second one.

## My setup
I have set up a prisma client in its own package which exposes a createContext function:

```javascript
import { PrismaClient } from '@prisma/client';
import type { Context } from './indextypes.d.js';

const prisma = new PrismaClient();

export const createContext = async (): Promise<Context> => ({ prisma });
```

I then import the package (using pnpm) into my Astro.build project and use the function to get access to my schema/client:

```javascript
import * as PrismaClient from 'my-package';
const { prisma } = await PrismaClient.createContext();
```

This has been working for me just fine up until the point at which I wanted to use two different databases with different schemas.
To achieve this, I adjusted my schema to generate into a different directory (i. e. `/src/generated-client`, but I have tried generating into `node_modules/@my-package/client` and others):

 ```javascript
 generator client {
  provider = "prisma-client-js"
  output   = "./src/generated-client"
}
```

After running `pnpm prisma generate` I was able to see a client being generated at the specified location, and pointed my imports to it:

```javascript
import { PrismaClient } from './src/generated-client';
import type { Context } from './indextypes.d.js';

const prisma = new PrismaClient();

export const createContext = async (): Promise<Context> => ({ prisma });
```

The other schema I simply went "the default way" in which I have a `prisma` directory with a `prisma.schema` within it. I generate this one using `pnpm prisma generate`. At this point, queries related to this schema work fine.

I re-generated the other client with `pnpm prisma generate` from the location of my-package. Now, all queries related to this package immediately fail with multiples of this issue:

```
[vite] Error when evaluating SSR module /home/user/projects/src/pages/api/v1/[item].json.js: failed to import "/@fs/home/user/projects/my-package/dist/index.js"
|- ReferenceError: exports is not defined
```

As this is when I started running into the aforementioned issue, I also tried generating the client from within the project I import my-package in using `pnpm prisma generate --schema ./node_modules/my-package/schema.prisma`, but to no avail.

TL;DR: How do I setup a (Astro.build) project using two different schemas with one schema coming from an installed package, and the other one being the "default" one?

I'm very sorry for the lack of a minimal reproducible example, I'm unsure how easy that would be to set up. Still, I would appreciate **any** help immensely! If needs be, I'll provide any further information that may help. :)

**Comment by ludralph** [comment]
Hi @Freakspot 👋 

Thank you for raising this question. I'm looking into this and will provide a response.

**Comment by Mxs2019** [comment]
Was anyone able to figure this out?

**Reply by jaycdev** [reply, marked as answer]
For me in SvelteKit, it seems like Vite (or pnpm, not sure which) does something special with node_module imports.

I moved my client's output into the **node_modules** folder (instead of my src directory) inside its own folder/module calling it **@prisma-something** and then referencing this generated client in my PrismaClient import. All seemed to start working afterward.

schema.prisma
```ts
generator client {
  provider = "prisma-client-js"
  output   = "../node_modules/@prisma-app/client"
}
```

prisma.ts
```ts
import { PrismaClient } from '@prisma-app/client';
export const prisma = new PrismaClient();
```

It's important that you call it @prisma-*, as naming it something else doesn't seem to work.

**Reply by ddm50** [reply]
This seems to help, but why this happens is very interesting. Why when it's named @prisma-something it works and anything else it doesn't?

Another edit, nope, this doesn't work.  exports is not defined is there as usual, it disappeared then reappeared after i went through some other issues.

**Reply by haristoreprops** [reply]
I'm also seeing problems on build while using the above solution

**Reply by gcb** [reply]
yall not having issues on deployment with this?

**Comment by reisr3** [comment]
@ludralph This isn't resolved - I'm still unable get the Prisma + Vite working together with multiple schema.prisma files. May we re-open? It's pretty shaky to accept the proposed solution without insight into why renaming the module might work.

**Reply by AmreeshTyagi** [reply]
Yes. above hack solution is not working. This should be reopen or linked to #5030 because it is coming due to no esm support of Prisma.

**Comment by AmreeshTyagi** [comment]
I came across this issue in past few years ago, but I managed to resolved it with remix. But now, I am using react-router (remix) with vite and after spending 2 more days, I am not able to fix this issue with prisma. Tried lot of configuration changes. But no luck. :( 

My project is huge and switching to drizzle would take a huge effort, but seems like I could have migrated to drizzle faster than resolving this issue.  Prisma team, please do the needful to fix ESM issues in Prisma. It is really painful for the developers who trust Prisma a lot.

![image](https://github.com/user-attachments/assets/13369cfb-e6a5-4b80-b949-99386473e9e6)

**Comment by ianhernandez** [comment]
running into this issue after following this guide: https://www.prisma.io/docs/guides/use-prisma-in-pnpm-workspaces 
+ with a react-router v7 project in `apps/`

**Reply by glerups-it** [reply]
This works for me when working with https://github.com/Shopify/shopify-app-template-remix as this is a TypeScript project.

Then I get TypeScript generated client instead in my Prima output folder

<img width="148" height="154" alt="2025-12-06_14h02_13" src="https://github.com/user-attachments/assets/cd337119-3f3f-4893-9722-31bbfc27af52" />

**Comment by Nizvan018** [comment]
In case this is useful, i solved this way (i'm using Remix v2.16):

```ts
generator client {
  provider = "prisma-client"
  output   = "../app/generated/prisma"
}
```

I just changed `prisma-client-js` to `prisma-client`, and let the output url that the documentation says

**Reply by BShowen** [reply]
This works, but [Prisma docs](https://www.prisma.io/docs/orm/more/releases#early-access) says not to use this in Production as it may not work as expected just yet. 

Your solution works for me but now I get a typescript error when doing 
```
import { Prisma } from "./prisma/client";
const myReturnType = Prisma.PromiseReturnType<typeof getUserWithPosts>
``` 

I get a typescript error `Namespace '"/Users/bshowen/app/prisma/client/client".Prisma' has no exported member 'PromiseReturnType'.t`

Im using React router 7.

**Reply by b-goldney** [reply]
this works. thanks.

**Reply by TimKochDev** [reply]
This works! I followed https://www.prisma.io/docs/guides/react-router-7 but got trapped by the VSCode Prisma extension telling me `prisma-client-js` was the only option to use as the provider. Only after going back step by step through the tutorial I found that mistake.

**Reply by sandyUni** [reply]
I get an error of ```"PrismaClient" is not exported by "app/generated/prisma/index-browser.js", imported by "src/preload/index.ts"."```
![image](https://github.com/user-attachments/assets/8221552a-5b06-425c-8cfd-edb17d94fd88)

**Reply by JintaoXIAO** [reply]
anyone can explain why this works?

**Comment by ArrayIterator** [comment]
Any info? Prisma Claim modern etc etc etc.
but not resolve this?
Support typescript not supported precompiled ts?
